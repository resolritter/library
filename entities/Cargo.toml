[package]
name = "entities"
version = "0.1.0"
authors = ["resolritter <resolritter@users.noreply.github.com>"]
edition = "2018"

[dependencies]
surf = "2.0"
serde = { version = "1", features = ["derive"] }
serde_json = "1"
