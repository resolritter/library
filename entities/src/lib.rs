mod macros;
#[macro_use]
pub mod book;
mod data;
pub mod user;

pub use crate::data::*;
