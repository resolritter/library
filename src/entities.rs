use bastion::prelude::BastionContext;
use serde::{Deserialize, Serialize};
use sqlx::PgPool;
use std::sync::Arc;

pub struct App {
    pub db_pool: &'static Arc<&'static PgPool>,
    pub bastion: &'static Arc<&'static BastionContext>,
}
impl std::fmt::Debug for App {
    fn fmt(&self, _f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        Ok(())
    }
}

#[derive(Debug)]
pub struct Config {
    pub db_url: String,
}

#[derive(Clone)]
pub struct ServerState {
    pub app: &'static Arc<&'static App>,
}

#[derive(Serialize, Deserialize, Debug)]
pub struct Book {
    pub id: i32,
    pub title: String,
}

#[derive(Deserialize, Debug)]
pub struct BookGet {
    pub title: String,
}

#[derive(Debug)]
pub struct BookGetMessage {
    pub channel: crossbeam_channel::Sender<Option<Book>>,
    pub payload: BookGet,
    pub app: &'static App,
}
