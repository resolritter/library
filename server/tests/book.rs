use entities::{book, user, BookLeaseByTitlePayload, UserCreationPayload};
use insta::assert_snapshot;
use stdext::function_name;
use surf::StatusCode;
use tempdir::TempDir;
use test_utils::{format::format_test_name, read_snapshot, spawn_test_program, SpawnedTest};

#[async_std::test]
async fn test_get() {
    let test_name = format_test_name(function_name!());
    let tmp_dir = TempDir::new(&test_name).unwrap();
    let SpawnedTest {
        server_addr,
        log_dir,
        process: _,
    } = &spawn_test_program(&tmp_dir);

    let mut get = book::get(&server_addr, "Rapunzel").await;
    assert_snapshot!(test_name, get.body_string().await.unwrap());

    assert_snapshot!(read_snapshot(&log_dir));
}

#[async_std::test]
async fn test_lease() {
    let test_name = format_test_name(function_name!());
    let tmp_dir = TempDir::new(&test_name).unwrap();
    let SpawnedTest {
        server_addr,
        log_dir,
        process: _,
    } = &spawn_test_program(&tmp_dir);

    const WHOLE_DAY: i64 = 86400;
    let (_, sample_user) = user::create(
        &server_addr,
        &UserCreationPayload {
            email: "user@user.com".to_string(),
            access_level: 0,
        },
    )
    .await;
    let payload = BookLeaseByTitlePayload {
        title: "Rapunzel".to_string(),
        lease_id_req: sample_user.email,
        lease_length: WHOLE_DAY,
    };

    // Lease a book for a whole day
    assert_snapshot!(
        format!("{}_good", test_name),
        book::borrow(&server_addr, &payload)
            .await
            .body_string()
            .await
            .unwrap()
    );

    let bad_borrow = book::do_borrow(&server_addr, &payload).await.unwrap();
    assert!(bad_borrow.status() == StatusCode::Forbidden);

    assert_snapshot!(read_snapshot(&log_dir));
}
