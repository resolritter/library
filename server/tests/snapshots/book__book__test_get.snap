---
source: server/tests/book.rs
expression: get.body_string().await.unwrap()
---
{"id":2,"lease_id":null,"lease_until":null,"title":"Rapunzel"}
